File: solution-cpp/003_longest_substring_without_repeating_characters.cpp
/*
	003. Longest Substring Without Repeating Characters
	@name: li jin
	@date: Feb 19, 2016
	@link: https://leetcode.com/problems/longest-substring-without-repeating-characters/
	@time: 16 ms
*/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int idx[256];
        for (int i = 0; i < 256; idx[i++] = -1);
        int ans = 0;
        int a, b;
        for (a = b = 0; b < s.length(); b++) {
            ans = max(ans, b - a);
            a = max(a, idx[s[b]] + 1);
            idx[s[b]] = b;
        }
        return max(ans, b - a);
    }
};

File: solution-python/003_longest_substring_without_repeating_characters.py
"""
	003. Longest Substring Without Repeating Characters
	@name: li jin
	@date: Feb 19, 2016
	@link: https://leetcode.com/problems/longest-substring-without-repeating-characters/
	@time: 88 ms
"""

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        idx = {}
        ans = a = b = 0
        for c in s:
            if c in idx and idx[c] >= a:
                ans = max(ans, b - a)
                a = idx[c] + 1
            idx[c] = b
            b += 1
        return max(ans, b - a)
